ndonnx/extensions.py

```diff
@@ -0,0 +1,437 @@
+# Copyright (c) QuantCo 2023-2025
+# SPDX-License-Identifier: BSD-3-Clause
+
+from __future__ import annotations
+
+from collections.abc import Mapping, Sequence
+from typing import Literal, TypeAlias, TypeVar, get_args
+from warnings import warn
+
+import numpy as np
+from typing_extensions import TypeIs
+
+import ndonnx as ndx
+
+from . import _typed_array as tydx
+from ._typed_array.masked_onnx import TyMaArray
+
+Scalar = TypeVar("Scalar", int, float, str)
+
+Key: TypeAlias = Scalar
+Value = TypeVar("Value", int, float, str)
+
+
+def shape(x: ndx.Array, /) -> ndx.Array:
+    """Returns shape of an array.
+
+    Parameters
+    ----------
+    x: Array
+        Array to get shape of
+
+    Returns
+    -------
+    out: Array
+        Array of shape
+    """
+    warn(
+        "'ndonnx.shape' is deprecated in favor of 'ndonnx.Array.dynamic_shape'",
+        DeprecationWarning,
+    )
+    return x.dynamic_shape
+
+
+def isin(x: ndx.Array, /, items: Sequence[Scalar]) -> ndx.Array:
+    """Return true where the input ``Array`` contains an element in ``items``.
+
+    ``NaN`` values do **not** compare equal.
+
+    Parameters
+    ----------
+    x: Array
+        The input Array to check for the presence of items.
+    items: Sequence[Scalar]
+        Scalar items to check for in the input Array.
+
+    Returns
+    -------
+    out: Array
+        Array of booleans indicating whether each element of ``x`` is in ``items``.
+    """
+    return ndx.Array._from_tyarray(x._tyarray.isin(items))
+
+
+# TODO: Bad naming: It is obvious from the type hints that this mapping is "static"
```

Comment:
```suggestion
```

I would dispute that it's a "bad" name to say the mapping is static, but in any case, we haven't been shy to break the API here so I would either remove the comment or change the name if you're keen on doing so.

Reply:
The breaking change in the public API is actually quite limited.

Reply:
I clarified the comment.